### PR TITLE
[dynamo] Dataclass variables with default field

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104840

The main complexity comes from the __init__ function of Dataclass variables which look something like this

```
[2023-07-10 05:01:29,548] torch._dynamo.symbolic_convert: [DEBUG] INLINING <code object __init__ at 0x7f7015154450, file "<string>", line 2>
  3           0 LOAD_FAST                1 (b)
              2 LOAD_FAST                0 (self)
              4 STORE_ATTR               0 (b)

  4           6 LOAD_FAST                2 (named_tensors)
              8 LOAD_DEREF               0 (_HAS_DEFAULT_FACTORY)
             10 IS_OP                    0
             12 POP_JUMP_IF_FALSE       20
             14 LOAD_DEREF               1 (_dflt_named_tensors)
             16 CALL_FUNCTION            0
             18 JUMP_FORWARD             2 (to 22)
        >>   20 LOAD_FAST                2 (named_tensors)
        >>   22 LOAD_FAST                0 (self)
             24 STORE_ATTR               1 (named_tensors)
             26 LOAD_CONST               0 (None)
             28 RETURN_VALUE
```

There are multiple issues
* VariableBuilder call in functions.py was wrong. We were calling *options as args.
* We were not setting source while tracking the new object. This led to no source for Dataclass variable, which has some new variables in its closures as seen in the above bytecode.
* There is IS_OP in above bytecode, which brings more cases.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
You probably didn't mean to push the cutlass change.

**Comment by anijain2305** [comment]
> You probably didn't mean to push the cutlass change.

Yes, by mistake. I think other parts also need more work. Putting WIP tag for now.

File: torch/_dynamo/variables/functions.py [hunk 1]
```diff
@@ -59,13 +62,13 @@ def wrap_bound_arg(tx, val, options, source=None):
     elif istensor(val):
         from torch._dynamo.variables.builder import VariableBuilder
 
-        return VariableBuilder(tx, source=source, **options)(val)
+        return VariableBuilder(tx, source=source)(val).add_options(options)
```
Comment (anijain2305) [inline torch/_dynamo/variables/functions.py:65]: VariableBuilder signature is (tx, source). I guess, earlier the options were always empty and therefore this bug stayed hidden.

File: test/dynamo/test_functions.py [hunk 2]
```diff
@@ -1248,6 +1249,72 @@ def func():
...
+            # Check that the default members are properly initialized
+            if isinstance(a.named_tensors, dict):
+                x = torch.sin(x)
+
+            # Change dataclass
+            c.scalar = 6
+            c.named_tensors["x"] = x
+
+            # Return dataclaass as well to check reconstruction
+            return c, torch.cos(x) * scaled_value + b.named_tensors["x"] + c.scalar
```
Comment (anijain2305) [inline test/dynamo/test_functions.py:1283]: @jansel the test returns a modified dataclass object and compares them later to validate the reconstruction logic.

**Comment by anijain2305** [comment]
@jansel This is ready for review.

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/builtin.py [hunk 3]
```diff
@@ -1299,6 +1300,16 @@ def _unimplemented():
         if isinstance(left, ConstantVariable) and isinstance(right, ConstantVariable):
             return ConstantVariable(op(left.value, right.value))
 
+        if isinstance(left, UserDefinedObjectVariable) and isinstance(
+            right, UserDefinedObjectVariable
+        ):
+            return ConstantVariable(op(left.value, right.value))
+
+        if op.__name__ == "is_":
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builtin.py:1308]: Unforunately, this is wrong. `SymNodeVariable(2) is ConstantVariable(2)` but their VariableTracker types are different.

**Pull request closed without merging**